ui-spacetimechart/src/components/SimpleRect.tsx

```diff
@@ -0,0 +1,93 @@
+  | ((ctx: CanvasRenderingContext2D) => string | CanvasPattern | CanvasGradient | null);
+
+export type SimpleRectProps = {
+  timeStart: Date;
+  timeEnd: Date;
+  spaceStart: number; // mm
+  spaceEnd: number; // mm
+  fillStyle: FillStyle;
+  border?: boolean;
+};
+
+export const SimpleRect = ({
+  timeStart,
+  timeEnd,
+  spaceStart,
+  spaceEnd,
+  fillStyle,
+  border,
+}: SimpleRectProps) => {
+  const overlayRef = useRef<HTMLDivElement>(null);
+  const drawRegion = useCallback<DrawingFunction>(
+    (ctx, { getSpacePixel, getTimePixel, spaceAxis }) => {
+      const timeStartPixel = getTimePixel(Number(timeStart));
+      const endTimePixel = getTimePixel(Number(timeEnd));
+      const spaceStartPixel = getSpacePixel(spaceStart);
+      const spaceEndPixel = getSpacePixel(spaceEnd);
+
+      const areaSpaceSize = spaceEndPixel - spaceStartPixel;
+      const areaTimeSize = endTimePixel - timeStartPixel;
+      if (!areaSpaceSize || !areaTimeSize) return;
+
+      ctx.save();
+      if (typeof fillStyle === 'function') {
+        const pattern = fillStyle(ctx);
+        if (!pattern) {
+          return;
+        }
+        ctx.fillStyle = pattern;
+      } else {
+        ctx.fillStyle = fillStyle;
+      }
+      if (spaceAxis === 'x') {
+        ctx.translate(spaceStartPixel, timeStartPixel);
+        ctx.fillRect(0, 0, areaSpaceSize, areaTimeSize);
+      } else {
+        ctx.translate(timeStartPixel, spaceStartPixel);
+        ctx.fillRect(0, 0, areaTimeSize, areaSpaceSize);
+      }
+      ctx.restore();
+      if (border && overlayRef.current) {
+        let borderSpaceSize, borderTimeSize, borderSpaceStart, borderTimeStart;
+        if (areaSpaceSize >= 0) {
+          borderSpaceSize = areaSpaceSize;
+          borderSpaceStart = spaceStartPixel;
+        } else {
+          borderSpaceSize = -areaSpaceSize;
+          borderSpaceStart = spaceStartPixel + areaSpaceSize;
+        }
+        if (areaTimeSize >= 0) {
+          borderTimeSize = areaTimeSize;
+          borderTimeStart = timeStartPixel;
+        } else {
+          borderTimeSize = -areaTimeSize;
+          borderTimeStart = timeStartPixel + areaTimeSize;
+        }
+        if (spaceAxis === 'x') {
+          overlayRef.current.style.left = `${borderSpaceStart}px`;
+          overlayRef.current.style.top = `${borderTimeStart}px`;
+          overlayRef.current.style.width = `${borderSpaceSize}px`;
+          overlayRef.current.style.height = `${borderTimeSize}px`;
+        } else {
+          overlayRef.current.style.left = `${borderTimeStart}px`;
+          overlayRef.current.style.top = `${borderSpaceStart}px`;
+          overlayRef.current.style.width = `${borderTimeSize}px`;
+          overlayRef.current.style.height = `${borderSpaceSize}px`;
+        }
+      }
+    },
+    [timeStart, timeEnd, spaceStart, spaceEnd, fillStyle, border]
+  );
+  useDraw('background', drawRegion);
+
+  return <div ref={overlayRef} className="simple-rect-border-box" />;
```

Comment:
Any reason why half of the rectangle (the background) is rendered with a canvas, and the other half (the border) is rendered with CSS? I think this is surprising in terms of z-position (the border will always be on top of everything else) and in terms of performance (the div will not behave well when instanced many times).

Reply:
I couldn’t get any satisfying rendering with the canvas, it’s very blurry.

Reply:
One needs to be careful wrt. pixel grid alignment when rendering lines on the canvas. See e.g. https://github.com/OpenRailAssociation/osrd-ui/pull/914

But I really don't think using mixed canvas and HTML is a good idea here.

Reply:
I retried canvas, seems much better even without any trying to align pixels.

Reply:
Nice!